atomic_reactor/plugins/pre_fetch_sources.py

```diff
@@ -0,0 +1,208 @@
+            nvr_build_id = nvr_build['build_id']
+
+        if self.koji_build_id and nvr_build_id and (self.koji_build_id != nvr_build_id):
+            err_msg = (
+                'koji_build_id {} does not match koji_build_nvr {} with id {}. '
+                'When specifying both an id and an nvr, they should point to the same image build'
+                .format(self.koji_build_id, self.koji_build_nvr, nvr_build_id)
+                )
+            raise ValueError(err_msg)
+        return self.koji_build_id or nvr_build_id
+
+    def assemble_srpm_url(self, base_url, srpm_filename, sign_key=None):
+        """Assemble the URL used to fetch an SRPM file
+
+        :param base_url: str, Koji root base URL with the given build artifacts
+        :param srpm_filename: str, name of the SRPM file
+        :param sign_key: str, key used to sign the SRPM, as listed in the signing intent
+        :return: list, strings with URLs pointing to SRPM files
+        """
+        url_components = [base_url, 'src', srpm_filename]
+        if sign_key:
+            url_components[1:1] = ['data', 'signed', sign_key]
+        return '/'.join(url_components)
+
+    def get_srpm_urls(self, build_id, sigkeys=None, insecure=False):
+        """Fetch SRPM download URLs for each image generated by a build
+
+        Build each possible SRPM URL and check if the URL is available,
+        respecting the signing intent preference order.
+
+        :param build_id: int, Koji build id of the container image we want SRPMs for
+        :param sigkeys: list, strings for keys which signed the srpms to be fetched
+        :return: list, strings with URLs pointing to SRPM files
+        """
+        srpm_urls = set()
+        archives = self.session.listArchives(build_id, type='image')
+        if not sigkeys:
+            sigkeys = ['']
+
+        for archive in archives:
+            rpms = self.session.listRPMs(imageID=archive['id'])
+            for rpm in rpms:
+                rpm_hdr = self.session.getRPMHeaders(rpm['id'], headers=['SOURCERPM'])
+                srpm_filename = rpm_hdr['SOURCERPM']
+                rpm_build = self.session.getBuild(rpm['build_id'])
+                path_info = get_koji_path_info(self.workflow, self.koji_fallback)
+                base_url = path_info.build(rpm_build)
+                for sigkey in sigkeys:
+                    url_candidate = self.assemble_srpm_url(base_url, srpm_filename, sigkey)
+                    req_session = get_retrying_requests_session()
+                    request = req_session.head(url_candidate, verify=not insecure)
+                    if request.status_code == requests.codes.ok:
+                        srpm_url = url_candidate
+                        break
+                try:
+                    srpm_urls.add(srpm_url)
+                except NameError:
+                    raise RuntimeError('Could not satisfy signing intent for {}'.format(rpm['nvr']))
```

Comment:
This doesn't look quite right. Can `set.add` fail? I wouldn't expect to see `NameError` here unless there's a typo in the source code.

Maybe `except NameError:` should really be `else:`, as in for...else...?

I think we're missing something like 'if we already found source for this n-v-r, move on to the next one', since most of the RPMs in each archive will be the same but for different architectures.

Reply:
Yes, relying on NameError here is no good: the next time round the rpms loop srpm_url will still be defined from the previous rpm.

Reply:
I agree this looks fragile, any future changes can break this assumption

Reply:
I added a dict to store srpm names and base urls to avoid making extra requests to koji for duplicated sources (multiple arches). Not relying on name error anymore and we will collect all sources that could not be fetched before halting to report all missing files.